File: .github/actions/test-system-io-ai-triage/src/replay.ts
/**
 * Replay — measure the classifier on runs TSIO has already ingested.
 *
 * The collection window (spec §7 step 1) merges the server and leaves the
 * calling repository's workflows alone, so no CI job ever calls this action and
 * no verdict is ever written. History accumulates; accuracy does not. That is
 * the one number blocking the gate.
 *
 * Replay closes it without a shadow flag and without a second code path that
 * only exists to be inert: it walks failed runs the server already holds and
 * puts each one through the SAME pieces a live run uses — `fetchEvidence`, the
 * deterministic classifier, `investigate`, `decide`, `writeLedger`. The rows it
 * writes are real. `check_state` and `waived` are real. Nothing flips, because
 * no CI job is listening.
 *
 * Two deliberate differences from a live run, both of which exist so the
 * measurement cannot be flattering:
 *
 *   - `mode` is always "gate". Replay is asking "what WOULD the gate have
 *     done", so waiving must be permitted; a shadow-mode replay would record
 *     `waived=false` everywhere and measure nothing.
 *   - the ledger batch is marked `replay: true`, so GET /triage/accuracy counts
 *     it separately. A replay verdict is decided with later runs of the same
 *     test already in the database — folding it into the live figure would
 *     overstate what CI actually does.
 *
 * No commit status is set and no PR comment is posted. Those are the two things
 * a developer would see, and a measurement must not reach into a PR that closed
 * weeks ago.
 */
import * as core from "@actions/core";

import { investigate } from "./agent.ts";
import {
  attachBlame,
  compareCommits,
  fetchEvidence,
  getPrDiff,
  getTestSource,
  listChangedFiles,
  writeLedger,
} from "./main.ts";
import { decide } from "./policy.ts";
import { retryFetch, parseJSON } from "./retry-fetch.ts";
import type { CompositeIdentity, Decision } from "./types.ts";

const PRODUCTION_URL = "https://test-system-io.internal.mattermost.com";
const STAGING_URL = "https://test-system-io-staging.internal.mattermost.com";

/** Cap per cluster, same as a live run: the model is the expensive part. */
const MAX_AGENT_CLUSTERS = 8;

interface ReplayCandidate {
  group_id: string;
  repository: string;
  branch: string;
  commit_sha: string;
  gh_run_id: string;
  gh_pr_number: number | null;
  name: string;
  failed: number;
}

export async function runReplay(): Promise<void> {
  const baseURL = core.getInput("use-staging") === "true" ? STAGING_URL : PRODUCTION_URL;
  const audience = core.getInput("oidc-audience") || "mattermost-test-system-io";
  const repo = core.getInput("replay-repo", { required: true });
  const baseline = core.getInput("baseline-branch") || "main";
  const branch = core.getInput("replay-branch");
  const days = clampInt(core.getInput("replay-days"), 30, 1, 180);
  const limit = clampInt(core.getInput("replay-limit"), 20, 1, 500);
  const githubToken = core.getInput("github-token");
  const anthropicKey = core.getInput("anthropic-api-key");
  const model = core.getInput("claude-model") || "claude-sonnet-4-6";

  if (!anthropicKey) {
    // Without a key the model never runs, every ambiguous cluster stays on its
    // history suggestion, and the batch measures the deterministic half only.
    // That is a legitimate run — it is what the classifier alone can decide —
    // but it must be stated, because the number it produces is not the AI's.
    core.warning(
      "no anthropic-api-key: replay will record deterministic verdicts only, " +
        "and the resulting accuracy figure is the classifier's, not the model's",
    );
  }

  const candidates = await fetchCandidates(baseURL, repo, branch, days, limit);
  core.info(`replay: ${candidates.length} candidate run(s) for ${repo} over ${days}d`);
  if (candidates.length === 0) {
    core.info("nothing to replay — every ingested failing run already has a ledger row");
    return;
  }

  let adjudicated = 0;
  let ledgerFailures = 0;

  for (const c of candidates) {
    const identity: CompositeIdentity = {
      repository: c.repository,
      branch: c.branch,
      commit_sha: c.commit_sha,
      gh_run_id: c.gh_run_id,
      gh_run_attempt: "1",
      name: c.name,
    };

    let pack;
    try {
      pack = await fetchEvidence(baseURL, identity, c.group_id, baseline);
    } catch (err) {
      // One unreadable run must not end the batch: the point of replay is
      // volume, and a group that 404s (deleted report, pruned artifact) is a
      // fact about that run, not a failure of the job.
      core.warning(`skipping ${c.group_id}: ${(err as Error).message}`);
      continue;
    }

    const runType = pack.group.gh_pr_number ? "PR" : "MAIN";
    const changedFiles = await listChangedFiles(
      githubToken,
      pack.group.repository,
      pack.group.gh_pr_number,
    );

    const decisions: Decision[] = [];
    for (const cluster of pack.clusters || []) {
      const recovered =
        cluster.representative.retry_count > 0 || cluster.representative.status === "flaky";
      const needsAI = cluster.suggested.needs_ai || recovered;
      let ai = undefined;
      if (needsAI && anthropicKey && decisions.length < MAX_AGENT_CLUSTERS) {
        try {
          ai = await investigate(cluster, {
            baseURL,
            apiKey: anthropicKey,
            model,
            group: pack.group,
            baselineBranch: baseline,
            changedFiles,
            compareCommits: (base, head) =>
              compareCommits(githubToken, pack.group.repository, base, head),
            getPrDiff: () => getPrDiff(githubToken, pack.group.repository, pack.group.gh_pr_number),
            getTestSource: (path, sha) =>
              getTestSource(githubToken, pack.group.repository, path, sha),
          });
        } catch (err) {
          core.warning(`agent failed on ${cluster.signature}: ${(err as Error).message}`);
        }
      }

      const d = decide({
        failure: cluster.representative,
        runType,
        branch: pack.group.branch || "",
        changedFiles,
        ai,
        // Always "gate": replay asks what the gate WOULD have done.
        mode: "gate",
      });
      d.member_count = cluster.member_count;
      decisions.push(await attachBlame(d, cluster, githubToken, pack.group.repository));
    }

    if (decisions.length === 0) continue;

    const ok = await writeLedger(baseURL, audience, pack, decisions, model, true);
    if (!ok) {
      // A live run refuses to flip a check when the ledger write fails. Replay
      // flips nothing, so there is nothing to refuse — but an unrecorded
      // verdict is an unmeasured one, so it is counted and reported.
      ledgerFailures++;
      core.warning(`ledger write failed for ${c.group_id} — this run stays unmeasured`);
      continue;
    }
    adjudicated++;
    const waived = decisions.filter((d) => d.waived).length;
    core.info(
      `${c.group_id} ${runType} ${c.commit_sha.slice(0, 8)}: ` +
        `${decisions.length} cluster(s), ${waived} would have been waived`,
    );
  }

  core.info(
    `replay complete: ${adjudicated}/${candidates.length} run(s) recorded` +
      (ledgerFailures > 0 ? `, ${ledgerFailures} ledger failure(s)` : ""),
  );
  core.setOutput("replayed", String(adjudicated));
  core.setOutput("ledger_failures", String(ledgerFailures));
  if (ledgerFailures > 0) {
    core.setFailed(`${ledgerFailures} run(s) could not be recorded`);
  }
}

async function fetchCandidates(
  baseURL: string,
  repo: string,
  branch: string,
  days: number,
  limit: number,
): Promise<ReplayCandidate[]> {
  const params = new URLSearchParams({
    repo,
    days: String(days),
    limit: String(limit),
  });
  if (branch) params.set("branch", branch);
  const res = await retryFetch(
    `${baseURL}/api/v1/triage/replay/candidates?${params.toString()}`,
    {},
    "triage/replay/candidates",
  );
  if (!res.ok) {
    throw new Error(`triage/replay/candidates HTTP ${res.status} ${await res.text()}`);
  }
  const body = await parseJSON<{ candidates?: ReplayCandidate[] }>(
    res,
    "triage/replay/candidates",
  );
  return body.candidates ?? [];
}

/** Parses an action input that must land inside a range, falling back rather
 * than failing: a bad `replay-days` should not lose the whole batch. */
export function clampInt(raw: string, dflt: number, min: number, max: number): number {
  const n = Number.parseInt(raw, 10);
  if (!Number.isInteger(n)) return dflt;
  if (n < min) return min;
  if (n > max) return max;
  return n;
}

File: apps/server/internal/api/triage/replay.go
package triage

// Replay candidates — the collection window's one moving part.
//
// While the calling repository's workflows are unmerged, nothing writes a
// verdict, so history accumulates but accuracy does not. The replay job closes
// that gap: it walks runs TSIO has already ingested, re-adjudicates each one
// through the same evidence pack, classifier, model and policy layer a live
// triage job would use, and records the result as a real ledger row marked
// `replay` (migration 000034). No CI job reads those rows, so nothing flips.
//
// This file is only the worklist. Everything that decides anything lives in
// the action, deliberately: a second implementation of the policy gate in Go
// would be a second thing to keep in sync, and the gate is the part that must
// never drift.

import (
	"net/http"
	"time"

	"github.com/mattermost/mattermost-test-system-io/apps/server/internal/api"
)

type replayCandidate struct {
	GroupID    string    `json:"group_id"`
	Repository string    `json:"repository"`
	Branch     string    `json:"branch"`
	CommitSHA  string    `json:"commit_sha"`
	GHRunID    string    `json:"gh_run_id"`
	GHPRNumber *int      `json:"gh_pr_number"`
	Name       string    `json:"name"`
	Failed     int       `json:"failed"`
	CreatedAt  time.Time `json:"created_at"`
}

// ReplayCandidates serves GET /api/v1/triage/replay/candidates?repo=&days=&limit=
// — ingested runs that have at least one failed test and no verdict yet.
//
// Public read, same posture as the other aggregate reads: it returns run
// coordinates the report list already exposes, and the replay job consults it
// before every batch.
func (h *Handlers) ReplayCandidates(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	repo := q.Get("repo")
	if repo == "" {
		api.WriteError(w, r, errRepoRequired())
		return
	}
	days := parseInt(q.Get("days"), 30)
	if days < 1 || days > 180 {
		api.WriteError(w, r, errRepoRequiredWith("days must be between 1 and 180"))
		return
	}
	limit := parseInt(q.Get("limit"), 50)
	if limit < 1 || limit > 500 {
		api.WriteError(w, r, errRepoRequiredWith("limit must be between 1 and 500"))
		return
	}
	branch := q.Get("branch") // empty = every branch

	// A run qualifies when it has a failed case and no ledger row for that
	// (repository, commit_sha, gh_run_id). Matching on the run rather than on
	// the cluster is deliberate: a partially-adjudicated run would otherwise
	// come back forever, and the verdict upsert is keyed per cluster anyway, so
	// re-running a whole group is idempotent.
	rows, err := h.Pool.Query(r.Context(), `
		SELECT g.id::text, g.repository, g.branch, g.commit_sha, g.gh_run_id,
		       g.gh_pr_number, g.name, count(*)::int AS failed, g.created_at
		FROM report_groups g
		JOIN reports rp    ON rp.report_group_id = g.id
		JOIN suites s      ON s.report_id = rp.id
		JOIN test_cases tc ON tc.suite_id = s.id
		WHERE (g.repository = $1 OR split_part(g.repository, '/', 2) = $1)
		  AND ($2 = '' OR g.branch = $2)
		  AND g.created_at >= now() - make_interval(days => $3)
		  AND tc.status = 'failed'
		  AND NOT EXISTS (
		      SELECT 1 FROM triage_verdicts v
		      WHERE v.repository = g.repository
		        AND v.commit_sha = g.commit_sha
		        AND v.gh_run_id  = g.gh_run_id
		  )
		GROUP BY g.id, g.repository, g.branch, g.commit_sha, g.gh_run_id,
		         g.gh_pr_number, g.name, g.created_at
		ORDER BY g.created_at DESC
		LIMIT $4
	`, repo, branch, days, limit)
	if err != nil {
		h.logError("replay candidates", err)
		api.WriteError(w, r, api.ErrInternal)
		return
	}
	defer rows.Close()

	out := []replayCandidate{}
	for rows.Next() {
		var c replayCandidate
		if err := rows.Scan(&c.GroupID, &c.Repository, &c.Branch, &c.CommitSHA,
			&c.GHRunID, &c.GHPRNumber, &c.Name, &c.Failed, &c.CreatedAt); err != nil {
			h.logError("replay candidates scan", err)
			api.WriteError(w, r, api.ErrInternal)
			return
		}
		out = append(out, c)
	}
	if err := rows.Err(); err != nil {
		h.logError("replay candidates rows", err)
		api.WriteError(w, r, api.ErrInternal)
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"repo":       repo,
		"branch":     branch,
		"days":       days,
		"candidates": out,
		"count":      len(out),
	})
}
